pkg/helm/controller/controller.go

```diff
@@ -71,6 +78,113 @@ func Add(mgr manager.Manager, options WatchOptions) error {
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			o := e.Object.(*unstructured.Unstructured)
+			log.V(1).Info("Reconciling due to dependent resource deletion", "name", o.GetName(), "namespace", o.GetNamespace(), "apiVersion", o.GroupVersionKind().GroupVersion(), "kind", o.GroupVersionKind().Kind)
+			return true
+		},
+
+		// Reconcile when a dependent resource is updated, so that it can
+		// be patched back to the resource managed by the Helm release, if
+		// necessary. Ignore updates that only change the status and
+		// resourceVersion.
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			old := e.ObjectOld.(*unstructured.Unstructured).DeepCopy()
+			new := e.ObjectNew.(*unstructured.Unstructured).DeepCopy()
+
+			delete(old.Object, "status")
+			delete(new.Object, "status")
+			old.SetResourceVersion("")
+			new.SetResourceVersion("")
+
+			if reflect.DeepEqual(old.Object, new.Object) {
+				return false
+			}
+			log.V(1).Info("Reconciling due to dependent resource update", "name", new.GetName(), "namespace", new.GetNamespace(), "apiVersion", new.GroupVersionKind().GroupVersion(), "kind", new.GroupVersionKind().Kind)
+			return true
+		},
+	}
+
+	var m sync.RWMutex
+	watches := map[schema.GroupVersionKind]struct{}{}
+	releaseHook := func(release *rpb.Release) error {
+		dec := yaml.NewDecoder(bytes.NewBufferString(release.GetManifest()))
+		for {
+			var u unstructured.Unstructured
+			err := dec.Decode(&u.Object)
+			if err == io.EOF {
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+
+			gvk := u.GroupVersionKind()
+			m.RLock()
+			_, ok := watches[gvk]
+			m.RUnlock()
+			if ok {
+				continue
+			}
+
+			restMapper := mgr.GetRESTMapper()
+			depMapping, err := restMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+			if err != nil {
+				return err
+			}
+			ownerMapping, err := restMapper.RESTMapping(owner.GroupVersionKind().GroupKind(), owner.GroupVersionKind().Version)
+			if err != nil {
+				return err
+			}
+
+			depClusterScoped := depMapping.Scope.Name() == meta.RESTScopeNameRoot
+			ownerClusterScoped := ownerMapping.Scope.Name() == meta.RESTScopeNameRoot
+
+			if !ownerClusterScoped && depClusterScoped {
+				m.Lock()
+				watches[gvk] = struct{}{}
+				m.Unlock()
+				log.Info("Cannot watch cluster-scoped dependent resource for namespace-scoped owner. Changes to this dependent resource type will not be reconciled",
+					"ownerApiVersion", r.GVK.GroupVersion(), "ownerKind", r.GVK.Kind, "apiVersion", gvk.GroupVersion(), "kind", gvk.Kind)
+				continue
+			}
+
+			err = c.Watch(&source.Kind{Type: &u}, &crthandler.EnqueueRequestForOwner{OwnerType: owner}, dependentPredicate)
+			if err != nil {
+				return err
+			}
+
+			m.Lock()
+			watches[gvk] = struct{}{}
+			m.Unlock()
+			log.Info("Watching dependent resource", "ownerApiVersion", r.GVK.GroupVersion(), "ownerKind", r.GVK.Kind, "apiVersion", gvk.GroupVersion(), "kind", gvk.Kind)
+		}
+	}
+	r.ReleaseHooks = append(r.ReleaseHooks, releaseHook)
```

Comment:
I'm probably missing something here but will there ever be more than 1 ReleaseHook per reconciler/controller?

Seems like we only need one releaseHook that parses the entire release manifest to add watches for all dependent resources to the controller.
So `r.ReleaseHooks` can just be a single `r.ReleaseHook`.

Reply:
There's not a use case for more than one right now. My thinking was that since `HelmOperatorReconciler` is exported, it's possible that users could roll their own logic in a release hook and add it themselves.

Maybe in the absence of such a request, we go with a single unexported release hook (i.e. `r.releaseHook`) for adding watches and only expand the public API if there's a specific request.

Reply:
@joelanford Yeah I would lean towards the latter for now since the intent is only to watch dependent resource for now. We can update this later if we get a use case.

Reply:
Sounds good :+1: